driver-async/src/main/com/mongodb/async/client/MongoCollectionImpl.java

```diff
@@ -180,37 +176,17 @@ public void count(final Object filter, final CountOptions options, final SingleR
     }
 
     @Override
-    public void distinct(final String fieldName, final Object filter, final SingleResultCallback<List<Object>> callback) {
-        distinct(fieldName, filter, new DistinctOptions(), callback);
+    public MongoIterable<Object> distinct(final String fieldName, final Object filter) {
+        return distinct(fieldName, filter, new DistinctOptions());
     }
 
     @Override
-    public void distinct(final String fieldName, final Object filter, final DistinctOptions options,
-                         final SingleResultCallback<List<Object>> callback) {
+    public MongoIterable<Object> distinct(final String fieldName, final Object filter, final DistinctOptions options) {
         DistinctOperation operation = new DistinctOperation(namespace, fieldName)
                                       .filter(asBson(filter))
                                       .maxTime(options.getMaxTime(MILLISECONDS), MILLISECONDS);
-        executor.execute(operation, readPreference, errorHandlingCallback(new SingleResultCallback<BsonArray>() {
-            @Override
-            public void onResult(final BsonArray result, final Throwable t) {
-                if (t != null) {
-                    callback.onResult(null, MongoException.fromThrowable(t));
-                } else {
-                    try {
-                        List<Object> distinctList = new ArrayList<Object>();
-                        for (BsonValue value : result) {
-                            BsonDocument bsonDocument = new BsonDocument("value", value);
-                            Document document = getCodec(Document.class).decode(new BsonDocumentReader(bsonDocument),
-                                    DecoderContext.builder().build());
-                            distinctList.add(document.get("value"));
-                        }
-                        callback.onResult(distinctList, null);
-                    } catch (Throwable tr) {
-                        callback.onResult(null, new MongoException("Error when decoding distinct results", tr));
-                    }
-                }
-            }
-        }));
+
+        return new OperationIterable<Object>(operation, readPreference, executor);
```

Comment:
Seems like this is a behavior change.  The previous implementation used the Document codec to transform each BsonValue into it's corresponding value type defined by the DocumentCodec.  Now I think it will return BsonValue.